mobile-widgets/uploadDiveLogsDE.cpp

```diff
@@ -0,0 +1,296 @@
+			continue;
+		}
+		xmlDocDumpMemory(transformed, (xmlChar **)&membuf, &streamsize);
+		xmlFreeDoc(doc);
+		xmlFreeDoc(transformed);
+		
+		/*
+		 * Save the XML document into a zip file.
+		 */
+		snprintf(xmlfilename, PATH_MAX, "%d.xml", i + 1);
+		s = zip_source_buffer(zip, membuf, streamsize, 1);
+		if (s) {
+			int64_t ret = zip_add(zip, xmlfilename, s);
+			if (ret == -1) {
+				qDebug() << errPrefix << "failed to include dive:" << i;
+			}
+		}
+	}
+	xsltFreeStylesheet(xslt);
+	if (zip_close(zip)) {
+		int ze, se;
+#if LIBZIP_VERSION_MAJOR >= 1
+		zip_error_t *error = zip_get_error(zip);
+		ze = zip_error_code_zip(error);
+		se = zip_error_code_system(error);
+#else
+		zip_error_get(zip, &ze, &se);
+#endif
+		report_error(qPrintable(tr("error writing zip file: %s zip error %d system error %d - %s")),
+					 qPrintable(QDir::toNativeSeparators(filename)), ze, se, zip_strerror(zip));
+		return false;
+	}
+	return true;
+}
+
+
+void uploadDiveLogsDE::uploadDives(const QString &filename, const QString &userid, const QString &password)
+{
+	QHttpPart part1, part2, part3;
+	static QNetworkRequest request;
+	QString args;
+
+
+	// Check if there is an earlier request open
+	if (reply != NULL && reply->isOpen()) {
+		reply->abort();
+		delete reply;
+		reply = NULL;
+	}
+	if (multipart != NULL) {
+		delete multipart;
+		multipart = NULL;
+	}
+	multipart = new QHttpMultiPart(QHttpMultiPart::FormDataType);
+
+	// prepare header with filename (of all dives) and pointer to file
+	args = "form-data; name=\"userfile\"; filename=\"" + filename + "\"";
+	part1.setRawHeader("Content-Disposition", args.toLatin1());
+	QFile *f = new QFile(filename);
+	if (!f->open(QIODevice::ReadOnly)) {
+		qDebug() << "ERROR opening zip file: " << filename;
+		return;
+	}
+	part1.setBodyDevice(f);
+	multipart->append(part1);
+
+	// Add userid
+	args = "form-data; name=\"user\"";
+	part2.setRawHeader("Content-Disposition", args.toLatin1());
+	part2.setBody(qPrefCloudStorage::divelogsde_userid().toUtf8());
+	multipart->append(part2);
+
+	// Add password
+	args = "form-data; name=\"pass\"";
+	part3.setRawHeader("Content-Disposition", args.toLatin1());
+	part3.setBody(qPrefCloudStorage::divelogsde_password().toUtf8());
+	multipart->append(part3);
+
+	// Prepare network request
+	request.setUrl(QUrl("https://divelogs.de/DivelogsDirectImport.php"));
+	request.setRawHeader("Accept", "text/xml, application/xml");
+	request.setRawHeader("User-Agent", getUserAgent().toUtf8());
+
+	// Execute async.
+	reply = manager()->post(request, multipart);
+
+	// connect signals from upload process
+	connect(reply, SIGNAL(finished()), this, SLOT(uploadFinished()));
+	connect(reply, SIGNAL(error(QNetworkReply::NetworkError)), this,
+		SLOT(uploadError(QNetworkReply::NetworkError)));
+	connect(reply, SIGNAL(uploadProgress(qint64, qint64)), this,
+		SLOT(updateProgress(qint64, qint64)));
+	connect(&timeout, SIGNAL(timeout()), this, SLOT(uploadTimeout()));
+
+	timeout.start(30000); // 30s
```

Comment:
this function also borrows heavily from the `DivelogsDeWebServices` class, only here you combined a rather complex multi function approach into something that is easier to read - I don't know if it's better, need to spend more time with the code, but again the real question is "why would it be worth to have this code twice"?

Reply:
Reason is that it is a multi version in the desktop is that the second part is triggered by the button, however I might be able to bypass that.